text/0000-semantic-private-in-public.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: semantic_private_in_public
+- Start Date: 2016-07-09
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Enforce that public APIs do not expose private definitions at the semantic level, while allowing the use of private aliases and blanket implementations for convenience and automation.
+
+# Motivation
+[motivation]: #motivation
+
+The "private-in-public" rules ensure the transitivity of abstraction. That is, one must be able to name types and bounds used in public APIs, from anywhere else.
+This property can be relied upon to create perfect proxies for generic types, functions and trait implementations.
+However, the current set of rules is too strict and ignores any semantic equivalence:
+```rust
+type MyResult<T> = Result<T, String>;
+
+#[derive(Clone)]
+struct Wrap<T>(T);
```

Comment:
Here's an example of leaking a private type through bounds, the returned type `T::Alias` seems to be completely public if checked separately. I wonder if something like this can be constructed with `PrivateWrap<T>: Bound`-like bounds. (I fully support the goals of the RFC, just trying to figure out possible holes.)
```
pub trait Tr<T> {
    type Alias = T;
}
impl<T, U> Tr<T> for U {}

struct Priv;
pub fn f<T>() -> T::Alias
    where T: Tr<Priv>
{
    panic!()
}
```